File: server-coreless/src/main/java/org/openqa/selenium/server/DefaultRemoteCommand.java
/*
 * Copyright 2006 ThoughtWorks, Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package org.openqa.selenium.server;

import java.io.*;
import java.net.*;
import java.util.*;

import org.openqa.selenium.server.browserlaunchers.*;

/**
 * The default implementation of the RemoteCommand interface
 * @author Paul Hammant
 * @version $Revision: 184 $
 */
public class DefaultRemoteCommand implements RemoteCommand {
    private final String command;
    private final String field;
    private final String value;
    private final String piggybackedJavaScript;

    public DefaultRemoteCommand(String command, String field, String value) {
        this.command = command;
        this.field = field;
        this.value = value;
        this.piggybackedJavaScript = null;
    }

    public DefaultRemoteCommand(String command, String field, String value, String piggybackedJavaScript) {
        this.command = command;
        this.field = field;
        this.value = value;
        if (piggybackedJavaScript!=null && !"".equals(piggybackedJavaScript)) {
            this.piggybackedJavaScript = piggybackedJavaScript;
        }
        else {
            this.piggybackedJavaScript = null;
        }
    }

    public String getCommandURLString() {
        return "cmd=" + LauncherUtils.urlEncode(command) + "&1=" + LauncherUtils.urlEncode(field) + "&2=" + LauncherUtils.urlEncode(value);
    }

    public String getCommand() {
        return command;
    }

    public String getField() {
        return field;
    }

    public String getValue() {
        return value;
    }

    public String toString() {
        if (piggybackedJavaScript==null) {
            return getCommandURLString();
        }
        return getCommandURLString() + "\n" + getPiggybackedJavaScript();
    }
    
    @Override public boolean equals(Object obj) {
        if (!(obj instanceof RemoteCommand)) return false;
        if (obj == null) return false;
        return this.toString().equals(obj.toString());
    };

    /** Factory method to create a RemoteCommand from a wiki-style input string */
    public static RemoteCommand parse(String inputLine) {
        if (inputLine == null) throw new NullPointerException("inputLine must not be null");
        if (!inputLine.startsWith("cmd=")) throw new IllegalArgumentException("invalid command string, missing 'cmd='=" + inputLine);
        

        String[] parts = inputLine.split("\\&");
        HashMap<String,String> args = new HashMap<String,String>();
        for (String part : parts) {
            String[] pair = part.split("\\=");
            if (pair.length < 2) {
                args.put(pair[0], "");
            } else {
                try {
                    args.put(pair[0], URLDecoder.decode(pair[1], "utf-8").trim());
                } catch (UnsupportedEncodingException e) {
                    throw new RuntimeException("Bug! utf-8 isn't supported???");
                }
            }
            
        }
        
        String command = args.get("cmd");
        String arg1 = args.get("1");
        String arg2 = args.get("2");
        return new DefaultRemoteCommand(command, arg1, arg2);
    }

    public String getPiggybackedJavaScript() {
        return piggybackedJavaScript;
    }
}

File: selenium/src/csharp/Core/DefaultRemoteCommand.cs
/*
 * Copyright 2006 ThoughtWorks, Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

using System;
using System.Web;
using System.Text;
namespace Selenium
{
	/// <summary>
	/// A representation of a single remote Command
	/// </summary>
	public class DefaultRemoteCommand : IRemoteCommand
	{
		private static readonly string PARSE_ERROR_MESSAGE = "Command string must contain 4 pipe characters and should start with a '|'. Unable to parse command string";
		private readonly string[] args;
		private readonly string command;

		/// <summary>
		/// Creates a command with the specified arguments
		/// </summary>
		/// <param name="command">the name of the command to run</param>
		/// <param name="args">its arguments (convert non-string arguments to strings)</param>
		public DefaultRemoteCommand(string command, string[] args)
		{
			this.command = command;
			this.args = args;
		}

		/// <summary>
		/// The string token that we'll send to the server
		/// </summary>
		public string CommandString
		{
			get
			{
				StringBuilder sb = new StringBuilder("cmd=");
				sb.Append(HttpUtility.UrlEncode(command));
				if (args == null) return sb.ToString();
				for (int i = 0; i < args.Length; i++)
				{
					sb.Append('&').Append((i+1).ToString()).Append('=').Append(HttpUtility.UrlEncode(args[i]));
				}
				return sb.ToString();
			}
		}
		
		/// <summary>
		/// The name of the Selenium command verb
		/// </summary>
		public string Command
		{
			get { return command; }
		}

		/// <summary>
		/// The array of arguments for this command
		/// </summary>
		public string[] Args
		{
			get { return args; }
		}

		/// <summary>
		/// Parses a "wiki-style" command string, like this: |type|q|Hello World|
		/// </summary>
		/// <param name="commandString">a wiki-style command string to parse</param>
		/// <returns>a Remote Command object that implements the command string</returns>
		public static DefaultRemoteCommand Parse(string commandString)
		{
			if (commandString == null || commandString.Trim().Length == 0 || !commandString.StartsWith("|"))
			{
				throw new ArgumentException(PARSE_ERROR_MESSAGE + "'" + commandString + "'.");
			}

			string[] commandArray = commandString.Split(new char[] { '|' });
			
			if (commandArray.Length != 5)
			{
				throw new ArgumentException(PARSE_ERROR_MESSAGE + "'" + commandString + "'.");
			}
			
			return new DefaultRemoteCommand(commandArray[1], new String[] {commandArray[2], commandArray[3]});
		}
	}
}
